client/tink_keyset_helper_test.go

```diff
@@ -99,3 +104,116 @@ func TestConvertTinkKeysetHandleToBytes(t *testing.T) {
 		t.Fatalf("when convert tink keyset handle to bytes, the keyset becomes invalid")
 	}
 }
+
+// getDummyKnoxVersionList is a helper for test. It returns a dummy knox version list for testing and a map from
+// Tink key ID to knox version ID. The data of each version is a Tink keyset in bytes that contains a single Tink
+// key. Argument counts decides how many versions are in this dummy veriosn list. Argument templateFunc decides
+// the type of created Tink keyset in each version.
+func getDummyKnoxVersionList(
+	counts int,
+	templateFunc func() *tinkpb.KeyTemplate,
+) (knox.KeyVersionList, map[uint32]uint64) {
+	var dummyVersionList knox.KeyVersionList
+	tinkKeyIDToKnoxVersionID := make(map[uint32]uint64)
+	// counts decide how many versions this dummy version list will have
+	for i := 0; i < counts; i++ {
+		// get a tink keyset in bytes that contains a fresh single key and the keyID is not duplicated
+		var keysetInbytes []byte
+		for {
+			keysetHandle, err := keyset.NewHandle(templateFunc())
+			if keysetHandle == nil || err != nil {
+				fatalf("cannot get tink keyset handle: %v", err)
+			}
+			_, ok := tinkKeyIDToKnoxVersionID[keysetHandle.KeysetInfo().PrimaryKeyId]
+			if !ok {
+				// This is a not duplicated Tink Key ID, use it to add a new Knox version
+				// To be noticed, index i is used as the dummy knox version ID
+				tinkKeyIDToKnoxVersionID[keysetHandle.KeysetInfo().PrimaryKeyId] = uint64(i)
+				bytesBuffer := new(bytes.Buffer)
+				writer := keyset.NewBinaryWriter(bytesBuffer)
+				err := insecurecleartextkeyset.Write(keysetHandle, writer)
+				if err != nil {
+					fatalf("cannot write tink keyset: %v", err)
+				}
+				keysetInbytes = bytesBuffer.Bytes()
+				break
+			}
+		}
+		// Add a new version to the dummy version list. Only one Primary version, all others are Active version.
+		var status knox.VersionStatus
+		if i == 0 {
+			status = knox.Primary
+		} else {
+			status = knox.Active
+		}
+		// To be noticed, index i is used as dummy knox version ID and dummy creation time.
+		dummyVersionList = append(dummyVersionList, knox.KeyVersion{
+			ID:           uint64(i),
+			Data:         keysetInbytes,
+			Status:       status,
+			CreationTime: int64(i),
+		})
+	}
+	return dummyVersionList, tinkKeyIDToKnoxVersionID
+}
+
+func TestAddNewTinkKeyset(t *testing.T) {
+	keyTemplate := aead.AES256GCMKeyTemplate
+	// create a dummy version list has one hunderd thousand Tink keys, this large number of Tink keys
```

Comment:
typo: "hundred"

Reply:
Thank you for your comment! I revised the code accordingly.